### DISCUSSION TITLE
Support for fragments or equivalents with latex equations in revealjs presentations

### DISCUSSION THREAD

**Discussion by rundel** [quarto-dev/quarto-cli#3224] [Feature Requests]
Currently there does not seem to be any way to introduce progressive elements / lines of a latex equation block. This is quite an important feature for teaching that is available via `\pause` and related tools in beamer.

It is possible in a hacky way to avoid this by splitting things into multiple latex equations but this then precludes the use of aligned equations which is a significant cost to pay. 

This may be a fundamental limitation imposed by mathjax / katex but seems worth at least looking into.

**Comment by jlgraves-ubc** [comment]
Does `/toggle` work for what you want?  

https://mathjax.github.io/MathJax-demos-web/toggle-steps.html.html

https://www.math.union.edu/~dpvc/talks/2013-01-11.jmm/tex-toggle.html

I don't think Quarto's MathJax currently supports this, but it might be an easy way to integrate this.

**Reply by rundel** [reply]
I was not familiar with these options and hadn't come across them in my googling - the [reveal steps variant](https://mathjax.github.io/MathJax-demos-web/reveal-steps.html) seems closest to the desired behavior but the implementation looks a bit clunky and javascript based, but possibly automateable.

The toggle option seems to involve having a ton of repeated latex code which seems less than ideal for most use cases (but is more flexible).

**Comment by thht** [comment]
After lots of fiddling and great help from @mcanouil, I found a very easy solution: Just add the `fragment` css class to part of the equation like this:

```latex
$$
\begin{align}
  (x+1)^2
    &= \class{fragment}{(x+1)(x+1)}            \\[3px]
    &\class{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
    &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
    &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
    &\class{fragment}{{} = x^2 + 2x + 1}
\end{align}
$$
```

And you need to put this in `_quarto.yml` to disable the assistive mode because otherwise the `fragment` classes get duplicated:

```yaml
format:
  revealjs:
    include-before: [ '<script type="text/x-mathjax-config">MathJax.Hub.Config({tex2jax: {enableAssistiveMml: false}});</script>']
```

**Reply by mcanouil** [reply]
Where the `fragment` class is duplicated exactly?
I tried the following, but did not see see this "issue".

```
---
format: revealjs
---

## Slide

\begin{align}
  (x+1)^2
    &= \class{fragment}{(x+1)(x+1)}            \\[3px]
    &\class{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
    &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
    &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
    &\class{fragment}{{} = x^2 + 2x + 1}
\end{align}
```

(and yes, the double dollar is not required depending on what environment you are using, also sometimes you actually need to not use them because it enables the math display environment which conflicts with some environment)

**Reply by thht** [reply]
Mathjax adds some hidden duplicates of all the elements in the dom with the same classes. You can replicate this behavior by putting an extra slide after the one with the math:

```
---
format: revealjs
---

## Slide

\begin{align}
  (x+1)^2
    &= \class{fragment}{(x+1)(x+1)}            \\[3px]
    &\class{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
    &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
    &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
    &\class{fragment}{{} = x^2 + 2x + 1}
\end{align}

## Another slide
```

Now you should see that revealjs does not continue to the second slide right after all fragments are shown. You need to press space 5 more times...

**Reply by mcanouil** [reply]
Indeed ;)
But, disabling assistive MML does not seem to solve the issue though, am I missing something?

**Reply by thht** [reply]
It does for me....

**Reply by thht** [reply]
You can also try by right-clicking in the math and disabling it there. Solved the problem immediatly for me...

**Reply by mcanouil** [reply]
I think the command is not correct (or the way to include it), because it does not change anything.
While disabling it manually in a browser works, but this is "saved" in browser cache.

**Reply by thht** [reply]
yes, you are right...

**Reply by mcanouil** [reply]
Weirdly, you have to be explicit ...

TL;DR: always add `html-math-method: mathjax` when modifying MathJax options/extensions/configs.

The following (slighty different syntax using `include-in-header`) works and disable the assistive MathML and you see that in the menu.

https://user-images.githubusercontent.com/8896044/212767760-ade28c0d-a2b0-4ef8-92ad-983a726e2552.mov

<details><summary>Code</summary>

```markdown
---
title: "Untitled"
format:
  revealjs:
    html-math-method: mathjax
    include-in-header:
      - text: |
          <script>
          MathJax = {
            options: {
              menuOptions: {
                settings: {
                  assistiveMml: false
                }
              }
            }
          };
          </script>
---

## Slide

\begin{align}
  (x+1)^2
  &= \class{fragment}{(x+1)(x+1)}            \\[3px]
  &\class{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
  &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
  &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
  &\class{fragment}{{} = x^2 + 2x + 1}
\end{align}

## new slide

test
```

</details>

On the other hand, if you write the same thing without specifying you are using MathJax for maths, it does not work.

https://user-images.githubusercontent.com/8896044/212767789-5dae3e13-3313-4f05-b56a-5ad7b543e7af.mov

<details><summary>Code</summary>

```markdown
---
title: "Untitled"
format:
  revealjs:
    include-in-header:
      - text: |
          <script>
          MathJax = {
            options: {
              menuOptions: {
                settings: {
                  assistiveMml: false
                }
              }
            }
          };
          </script>
---

## Slide

\begin{align}
  (x+1)^2
  &= \class{fragment}{(x+1)(x+1)}            \\[3px]
  &\class{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
  &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
  &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
  &\class{fragment}{{} = x^2 + 2x + 1}
\end{align}

## new slide

test
```

</details>

**Reply by mcanouil** [reply]
In addition and for your information, KaTeX also has that ability with `\htmlClass`, see <https://katex.org/docs/support_table.html>.
<img width="794" alt="image" src="https://user-images.githubusercontent.com/8896044/212768571-802e95cb-bd23-497e-bbfa-d5e1e07be41c.png">

**Reply by thht** [reply]
But this cannot be enabled, because you need to set `trust = true` and `strict = false` when calling `katex.render` and there does not seem to be a way to add these options right now because they are hardcoded here: https://github.com/quarto-dev/quarto-cli/blob/2dbf31994a7b6da31c3de78da31e6d2db62c0ce3/src/command/render/template.ts#L244

**Reply by mcanouil** [reply]
Hum, I have to admit, I did not try the KaTeX equivalent.

Note that you can call several times the configurations function as in MathJax, so it should be possible.
I'll try later.

**Reply by cderv** [reply]
> Although, default options could be allowed to be modified, unless there is a reason. 🤔

Probably a feature request to make so that we can see how a customization could be offered.

**Reply by mcanouil** [reply]
@cderv I edited my answer because it might not need to add anything, instead just a matter of how to pass/call the configuration function.

**Reply by thht** [reply]
Another option is to not use quarto but revealjs to render the math:

1. Create a folder for the math plugin.
1.1. Copy all the files there
1.2. Create a `plugin.yml` file with the following content:
     ```yml
     name: RevealMath.KaTeX
     script: 'math.js'
     config:
       katex:
         strict: false
         trust: true
     ```
2. Activate the plugin in `_quarto.yml` (assuming the plugin folder is `plugins/math_plugin`:
  ```yml
    revealjs-plugins:
      - plugins/math_plugin
  ```
3. Put the math code in a ```` ```{=html} ```` fence:
  ````
  ```{=html}
  \[\begin{align}
    (x+1)^2
      &= \htmlClass{fragment}{(x+1)(x+1)}            \\[3px]
      &\htmlClass{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
      &\htmlClass{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
      &\htmlClass{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
      &\htmlClass{fragment}{{} = x^2 + 2x + 1}
  \end{align} \]
```
  ````

**Reply by fbob** [reply]
Do you think the solution proposed by @mcanouil and @thht could be proposed as a pull request for quarto ?
It's a very handy solution and this math fragment feature is great when doing classes.

**Reply by mcanouil** [reply]
The issue is that it depends on the Math framework used.
MathJax is the default, but Quarto allows to change to few others.
Thus, it can't really be implemented as a part of Quarto CLI, instead use the tricks showed here.

**Reply by fbob** [reply]
I understand the issue, thanks for the precision.

**Reply by RoyiAvital** [reply]
This is great!

2 questions:

1. Why there is a need for `{}` before `=` in `&\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]`?
2. Why is the `&` outside the class?

I am asking as I usually use something like:

```qmd
\begin{align}
  (x+1)^2
  &= \class{fragment}{(x+1)(x+1) && \text{some explanations}}            \\[3px]
  &\class{fragment}{{} = x(x+1) + 1(x+1) && \text{more explanations} }    \\[3px]
  &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
  &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
  &\class{fragment}{{} = x^2 + 2x + 1}
\end{align}
```

Yet it fails for me.

**Reply by ninaRjoyce** [reply]
The solution below works, but it seems to interfere some underlying process (I am new to Quarto and reveal.js, so please forgive the technical term) because, once all of the fragments are displayed, in order to advance to the next slide I have to click as many times as there are class{fragment} statements. For instance, in the example below I would click five times to move through the fragments and then I would have to click five more times to advance to the next slide. 

The other issue is that if I have any :::{fragment} divisions on my slide, these get rendered first, then the ones using \class{fragment}, and I can't control this with fragment-index.




> After lots of fiddling and great help from @mcanouil, I found a very easy solution: Just add the `fragment` css class to part of the equation like this:
> 
> ```
> $$
> \begin{align}
>   (x+1)^2
>     &= \class{fragment}{(x+1)(x+1)}            \\[3px]
>     &\class{fragment}{{} = x(x+1) + 1(x+1)}    \\[3px]
>     &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
>     &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
>     &\class{fragment}{{} = x^2 + 2x + 1}
> \end{align}
> $$
> ```
> 
> And you need to put this in `_quarto.yml` to disable the assistive mode because otherwise the `fragment` classes get duplicated:
> 
> ```yaml
> format:
>   revealjs:
>     include-before: [ '<script type="text/x-mathjax-config">MathJax.Hub.Config({tex2jax: {enableAssistiveMml: false}});</script>']
> ```

**Reply by rben01** [reply]
In my case I had to right click the rendered LaTeX to bring up the MathJax window, then disable “include hidden MathML”. For some reason I could not get this to work automatically, even with the updated MathJax 3 syntax,

```javascript
<script>
	window.MathJax = {
		options: {
			enableAssistiveMml: false,
		},
	};
</script>
```

**Reply by rben01** [reply]
> This is great!
> 
> 2 questions:
> 
> 1. Why there is a need for `{}` before `=` in `&\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]`?
> 2. Why is the `&` outside the class?
> 
> I am asking as I usually use something like:
> 
> ```
> \begin{align}
>   (x+1)^2
>   &= \class{fragment}{(x+1)(x+1) && \text{some explanations}}            \\[3px]
>   &\class{fragment}{{} = x(x+1) + 1(x+1) && \text{more explanations} }    \\[3px]
>   &\class{fragment}{{} = (x^2+x) + (x+1)}    \\[3px]
>   &\class{fragment}{{} = x^2 + (x + x) + 1}  \\[3px]
>   &\class{fragment}{{} = x^2 + 2x + 1}
> \end{align}
> ```
> 
> Yet it fails for me.

{} is needed before the = so that latex treats = as a binary operator. Without {}, you won't get proper spacing around the =.

& goes outside the fragment because it's not displayed per se, but is merely controlling latex alignment. There is nothing to “uncover” with &.

**Reply by mcanouil** [reply]
@rben01 would you mind showing a full but small reproducible document? This would help future readers.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

**Reply by rschmehl** [reply]
> The other issue is that if I have any :::{fragment} divisions on my slide, these get rendered first, then the ones using \class{fragment}, and I can't control this with fragment-index.

You might want to look at this fix (for Reveal.js running with MathJax4):
https://github.com/hakimel/reveal.js/pull/3811#issuecomment-3915630932

**Comment by willov** [comment]
Running on Zen browser, I couldn't get the include-before to disable the extra clicks needed. By default, the math renderer was set to SVG, and switch to be MathML made it  work. 

Alternatively, keeping with SVG and manually turning off Assistive MathML from the menu also worked, as mentioned by @rben01.
